### PR TITLE
LG-4319: Add expand/collapse prop to code component in React

### PR DESCRIPTION
**Pull request by tsck**
## ✍️ Proposed changes

🎟 _Jira ticket:_ [LG-4319](https://jira.mongodb.org/browse/LG-4319)

### For bug fixes, new features & breaking changes

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

### For new components

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README
- [x] I have verified the Live Example looks as intended on the design website.

## 🧪 How to test changes
- Unit tests
- Check `Code` for unexpected regressions

### CODE REVIEW

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [inline packages/code/src/Code/Code.styles.ts:127]: Should there be a transition?

File: packages/code/src/Code/Code.spec.tsx
Comment (shaneeza) [inline packages/code/src/Code/Code.spec.tsx:258]: We should use a testid here. This would also make it easier for consumers to target this button.

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [inline packages/code/src/Code/Code.styles.ts:132]: Will this always be 5?

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [inline packages/code/src/Code/Code.styles.ts:170]: Should this font size change when the `baseFontSize` changes?

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [inline packages/code/src/Code/Code.styles.ts:157]: Can we add transitions to this button?

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [inline packages/code/src/Code/Code.styles.ts:171]: Whenever possible, can you use spacing tokens?

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts:127]: I didn't see anything communicated about a transition on expansion/collapse. I've reached out to Rob to get his thoughts.

File: packages/code/src/Code.stories.tsx
Comment (tsck) [reply packages/code/src/Code.stories.tsx:1]: Will do!

File: packages/code/src/Code/Code.spec.tsx
Comment (tsck) [reply packages/code/src/Code/Code.spec.tsx:258]: Good point! I have to get into the design system mindset. I generally try to avoid test ID's to avoid bloating markup for test purposes, but what you're saying makes sense given how consumers might use it. Will update!

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts:132]: Yup. If there are less than 5 lines of code there won't be an expand/collapse button, even if `expandable` is `true`. So the collapsed version will always be 5 lines when visible.

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts:170]: Good catch! I would think that it should. I'll update this to include that.

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts:171]: Oh absolutely, will update shortly

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [inline packages/code/src/Code/Code.styles.ts:190]: Using `focus-visible` instead of `focus` is preferable here, as it prevents focus styles from appearing on mouse clicks.

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts:157]: Is there a transition that we typically use for buttons? I'll have a look and see if I can find something, but if there's a standard here that you could share, that would be helpful.

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [inline packages/code/src/Code/Code.styles.ts:191]: Would any of our color tokens work here? https://github.com/mongodb/leafygreen-ui/blob/main/packages/tokens/src/color/color.ts

File: packages/code/src/Code/Code.tsx
Comment (shaneeza) [inline packages/code/src/Code/Code.tsx:134]: Is it possible to use a `ref` here instead of an `id`?

File: packages/code/src/Code/Code.tsx
Comment (shaneeza) [inline packages/code/src/Code/Code.tsx:9]: If we add transitions, then we could mimic Stephen's changes in #2443

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts:190]: Just caught this with Rob actually! Will do

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts:127]: Discussed with Rob and will be adding a 300-500ms animation on expansion/collapse here

File: packages/code/src/Code/Code.tsx
Comment (tsck) [reply packages/code/src/Code/Code.tsx:9]: I like it, will do!

File: packages/code/src/Code/Code.tsx
Comment (shaneeza) [inline packages/code/src/Code/Code.tsx:191]: not: since this will always be 5, maybe we should make this a global constant since you're also using it in styles?

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [reply packages/code/src/Code/Code.styles.ts:132]: Can you add this to the tsdoc and README?

Comment (shaneeza) [reply packages/code/src/Code/Code.styles.ts:132]: Also, remember to add a changeset

File: packages/code/src/Code/Code.tsx
Comment (shaneeza) [inline packages/code/src/Code/Code.tsx:263]: nit: you could break this out into a new component.

**Review by shaneeza — CHANGES_REQUESTED** [review]
Looks good so far! Remember to update the README and add a changeset.

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts:191]: I was able to find one for the background, but I'm not sure I'm seeing one for the color. Feel free to let me know if I'm missing that though.

File: packages/code/src/Code/Code.tsx
Comment (tsck) [reply packages/code/src/Code/Code.tsx:134]: Yup, definitely is, and it's helpful for calculating the height now that we'll be transitioning. Great point!

File: packages/code/src/Code/Code.tsx
Comment (tsck) [reply packages/code/src/Code/Code.tsx:263]: I considered this, but really there wasn't much logic going on here so it didn't seem very helpful. If we decide to rotate the chevron, there may be more that will justify this though. I'll update shortly when I have more thoughts 🙂

File: packages/code/src/Code/Code.tsx
Comment (tsck) [reply packages/code/src/Code/Code.tsx:9]: I'm not sure the usage here is quite the same. This is primarily because both the text and the icon change on expansion/collapse. So rotating the icon alone doesn't inherently make the transition look less jarring. I have an idea and want to play around with a few possible solutions. I'll push those on separate commits before requesting another review, and probably run them by design to see what looks best here.

File: packages/code/src/Code/Code.tsx
Comment (tsck) [reply packages/code/src/Code/Code.tsx:193]: This makes sense to m. Could you point me to where we store global constants. I see `lib/` seems to be similar but I'm not quite sure that's correct.

Comment (shaneeza) [reply packages/code/src/Code/Code.tsx:195]: You'll notice that some components have a `constants.ts` file — you can add a constants.ts file inside `code/src`.

File: packages/code/src/Code.stories.tsx
Comment (tsck) [reply packages/code/src/Code.stories.tsx:1]: I chose not to do this because we don't follow the same pattern for `copyable`, which feels like it would be a similar use case here. If this is something we should still do please let me know, happy to discuss!

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts:134]: This is a really good point! I don't think there is! Let me remove the scrollbar. I'll have Rob give another look anyway before merging but I think this is a good catch.

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [reply packages/code/src/Code/Code.styles.ts:153]: Was there an issue with adding transitions?

File: packages/code/README.md
Comment (shaneeza) [inline packages/code/README.md:118]: Can you add that this will happen if there are more than five lines of code?

File: packages/code/src/Code/Code.tsx
Comment (shaneeza) [inline packages/code/src/Code/Code.tsx:51]: nit: you could create a `utils` folder inside `code/src` and add this there. `Combobox` is a good example.

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts:153]: Oops, nope! I think I just accidentally removed it when I removed the transition for the icon rotation. Will add it back now.

File: packages/code/src/types.ts
Comment (shaneeza) [inline packages/code/src/types.ts:27]: Instead of `HTMLElementProps`, could you use `ComponentPropsWithRef`?

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts]: Just a note - Rob has indicated that the button `font-size` should always be `13px`. Will be changing this back shortly.

File: packages/code/src/types.ts
Comment (shaneeza) [inline packages/code/src/types.ts:118]: Could you add that this only works if there are 5 or more lines of code?

File: packages/code/src/Code/Code.tsx
Comment (tsck) [reply packages/code/src/Code/Code.tsx:50]: So the goal in my mind when moving this into a named function wasn't so much to make it reusable, but rather to make it clear what `element.offsetHeight - element.clientHeight` is calculating. I'm not sure if this is a utility method that will be used elsewhere, so I think there's at least some benefit to keeping it close to its usage. Happy to discuss further if this is a well established pattern we follow!

File: .changeset/fresh-comics-thank.md
Comment (shaneeza) [inline .changeset/fresh-comics-thank.md:5]: Can you add more details on what the expandable prop does and link the JIRA ticket? [Here's an example of that.](https://github.com/mongodb/leafygreen-ui/blob/main/packages/tabs/CHANGELOG.md#1300)

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [inline packages/code/src/Code/Code.styles.ts:166]: let's use a [token](https://github.com/mongodb/leafygreen-ui/blob/main/packages/tokens/src/typeScales.ts#L1) here

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [inline packages/code/src/Code/Code.styles.ts:171]: Whats the z-index for?

File: packages/code/src/Code/Code.tsx
Comment (shaneeza) [reply packages/code/src/Code/Code.tsx:50]: I think it's okay that it's not reusable; it's more to avoid bloating the file, but I don't feel too strongly about this.

File: packages/code/src/Code/Code.styles.ts
Comment (shaneeza) [reply packages/code/src/Code/Code.styles.ts:127]: From what I've seen, we usually wait until a feature is fully supported before we use it, so I would hold off on using `auto`

File: packages/code/src/Code/Code.styles.ts
Comment (tsck) [reply packages/code/src/Code/Code.styles.ts:171]: It moves the button above the shadows. You'll see this done throughout the component in other styles. I've added a comment to indicate this.

File: packages/code/src/Code/Code.spec.tsx
Comment (shaneeza) [inline packages/code/src/Code/Code.spec.tsx:256]: I forgot to mention this earlier, but we recently decided to follow a [BEM-ish pattern](https://github.com/mongodb/leafygreen-ui/blob/main/STYLEGUIDE.md#follow-bem-ish-patterns-when-hard-coding-a-data-testid-or-data-lgid) for testids. Following the pattern, we can change this to: `lg-code-expand_button`.

File: packages/code/src/types.ts
Comment (shaneeza) [inline packages/code/src/types.ts:57]: You can remove this line since you're using `ComponentPropsWithRef`

File: packages/code/src/Code/Code.tsx
Comment (tsck) [reply packages/code/src/Code/Code.tsx:50]: I can see the trade-off here. Given that it's just a one line function created for readability, and is similar with the other util method above it, I don't love the idea of putting it further away from the usage, causing navigation between files to be necessary in order to gain context. I think I'd see benefit in just in-lining it in that case to be more in line with the [Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))

File: packages/code/src/types.ts
Comment (tsck) [reply packages/code/src/types.ts:57]: On second thought, this is no longer needed at all. I was originally adding a `ref` to access a `Sytnax` child, but found a way around this by using another reference. I'm going to revert these changes entirely.

**Review by audroue — APPROVED** [review]
LGTM!

**Review by shaneeza — APPROVED** [review]
✨

**Pull request merged**
